### PR TITLE
Register deployments with the portal.

### PR DESCRIPTION
**Pull request by cdonati**

### CODE REVIEW

File: lib/appscale.py [hunk 1]
```diff
@@ -77,6 +79,7 @@ class AppScale():
   destroy: Gracefully terminates the currently running AppScale deployment.
   down: An alias for 'destroy'.
   clean: Forcefully terminates all services in a cluster AppScale deployment.
+  register: Registers an AppScale deployment with the portal.
```
Comment (menivaitsi) [inline lib/appscale.py]: Change "portal" into "AppScale Portal".

File: lib/appscale.py [hunk 2]
```diff
@@ -131,6 +134,16 @@ def get_locations_json_file(self, keyname):
     return json_file
 
 
+  def get_nodes(self, keyname):
+    """Retrieve a list of the running nodes."""
```
Comment (menivaitsi) [inline lib/appscale.py]: Put spaces before and after """.

Comment (menivaitsi) [inline lib/appscale.py]: Add sections "Args" and "Returns" in the docstring.

File: lib/appscale.py [hunk 3]
```diff
@@ -131,6 +134,16 @@ def get_locations_json_file(self, keyname):
...
+    try:
+      with open(self.get_locations_json_file(keyname)) as f:
+        return json.loads(f.read())
+    except IOError:
+      raise AppScaleException("AppScale does not currently appear to"
+                              " be running. Please start it and try again.")
```
Comment (menivaitsi) [inline lib/appscale.py]: Indentation should be two spaces.

File: lib/appscale.py [hunk 4]
```diff
@@ -748,3 +753,47 @@ def clean(self):
     LocalState.cleanup_appscale_files(keyname)
     AppScaleLogger.success("Successfully shut down your AppScale deployment.")
     return all_ips
+
+
+  def register(self):
+    """'register' allows users to register their AppScale deployment with the
+    portal.
```
Comment (menivaitsi) [inline lib/appscale.py]: portal -> AppScale Portal

File: lib/appscale.py [hunk 5]
```diff
@@ -748,3 +753,47 @@ def clean(self):
...
+      'deployment_type': deployment_type,
+      'ip_address': public_ips,
+      'secret': secret
+    }
+    response = opener.open(RegistrationHelper.ADD_DEPLOYMENT_URL,
+                           urllib.urlencode(deployment_data, True))
+    deployment = json.loads(response.read())
+    url = RegistrationHelper.get_deployment_url(deployment['safe_name'])
+    print('Your AppScale deployment {0} has been added to the portal.\n'
+          'You can view it here: {1}'.format(name, url))
```
Comment (menivaitsi) [inline lib/appscale.py]: Fix indentation.

File: lib/appscale.py (hunk 4 above)
Comment (menivaitsi) [inline lib/appscale.py]: Make that change throughout.

File: lib/appscale.py (hunk 2 above)
Comment (menivaitsi) [inline lib/appscale.py]: Make these changes in all files changed.

File: lib/registration_helper.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+      if deployment['secret'] == secret:
+        deployment_url = cls.get_deployment_url(deployment['safe_name'])
+        print('This deployment has already been registered as {0}.\n'
+              'You can view it here: {1}'
+              .format(deployment['name'], deployment_url))
+        exit()
+
+  @classmethod
+  def login(cls):
+    username = raw_input('Portal Login Email: ')
```
Comment (menivaitsi) [inline lib/registration_helper.py]: Missing docstring.

File: lib/registration_helper.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    login_data = {'username': username, 'password': password, 'tools': True}
+    cookies = CookieJar()
+    opener = urllib2.build_opener(urllib2.HTTPCookieProcessor(cookies))
+    opener.addheaders = [('Referer', RegistrationHelper.LOGIN_URL)]
+
+    try:
+      opener.open(RegistrationHelper.LOGIN_URL, urllib.urlencode(login_data))
+      return opener
+    except urllib2.HTTPError as error:
+      if error.code == 401:
```
Comment (menivaitsi) [inline lib/registration_helper.py]: Define these error codes as constants at the top of the file.

File: lib/appscale.py [hunk 8]
```diff
@@ -131,6 +134,24 @@ def get_locations_json_file(self, keyname):
     return json_file
 
 
+  def get_nodes(self, keyname):
+    """Retrieve a list of the running nodes.
```
Comment (menivaitsi) [inline lib/appscale.py]: Put a whitespace after """ and before the text.
Do the same in the rest of the docstrings.

File: lib/appcontroller_client.py [hunk 9]
```diff
@@ -387,3 +387,25 @@ def set_property(self, property_name, property_value):
...
+
+
+  def deployment_id_exists(self):
+    """ Asks the AppController if the deployment ID is stored in ZooKeeper.
+
+    Returns:
+      A boolean indicating whether the deployment ID is stored or not.
+    """
+    return self.run_with_timeout(10,
+      'Check for deployment id request timed out.',
```
Comment (menivaitsi) [inline lib/appcontroller_client.py]: Make "id" upper case. Be consistent in all places too :)

File: lib/appscale.py [hunk 10]
```diff
@@ -131,8 +133,40 @@ def get_locations_json_file(self, keyname):
...
 
+  def get_nodes(self, keyname):
+    """ Retrieve a list of the running nodes.
+
+    Args:
+      keyname: An identifier for the AppScale deployment.
+    Returns:
+      A list of nodes in the running AppScale deployment.
+    Raises:
+      AppScaleException: If there is no locations json file.
```
Comment (menivaitsi) [inline lib/appscale.py]: json -> JSON

File: lib/appscale.py [hunk 11]
```diff
@@ -131,8 +133,40 @@ def get_locations_json_file(self, keyname):
...
+    """
+    try:
+      with open(self.get_locations_json_file(keyname)) as f:
```
Comment (menivaitsi) [inline lib/appscale.py]: Pylint doesn't like variable names like 'f'.
Please fix :)

File: lib/appscale.py [hunk 12]
```diff
@@ -131,8 +133,40 @@ def get_locations_json_file(self, keyname):
...
+      raise AppScaleException("AppScale does not currently appear to"
+        " be running. Please start it and try again.")
+
+
+  def get_head_node(self, nodes):
+    """ Retrieve a node with the 'shadow' role.
+
+    Args:
+      nodes: A list of nodes in the running AppScale deployment.
+    Returns: A string containing the IP address of the head node.
```
Comment (menivaitsi) [inline lib/appscale.py]: Put a line break after 'Returns:'.

File: lib/appscale.py [hunk 13]
```diff
@@ -748,3 +774,43 @@ def clean(self):
...
+    Portal.
+
+    Raises:
+      AppScaleException: If the deployment has already been registered.
+    """
+    appscale_yaml = yaml.safe_load(self.read_appscalefile())
+    if 'keyname' in appscale_yaml:
+      keyname = appscale_yaml['keyname']
+    else:
+      keyname = "appscale"
```
Comment (menivaitsi) [inline lib/appscale.py]: Use single quotes for consistency?

File: lib/registration_helper.py [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+      .read()
+    name = raw_input('Deployment Name [{0}]: '.format(default_name)).strip()
+    if name == '':
+      name = default_name
+
+    return name
+
+  @classmethod
+  def get_deployment_url(cls, safe_name):
+    """ Generate a url that opens the deployment directly.
```
Comment (menivaitsi) [inline lib/registration_helper.py]: url => URL

File: lib/registration_helper.py [hunk 15]
```diff
@@ -0,0 +1,211 @@
...
+
+    Args:
+      safe_name: A string containing a version of the deployment nickname
+        that's safe for use as an HTML ID attribute.
+    Returns: A string containing a link to the deployment on the portal.
```
Comment (menivaitsi) [inline lib/registration_helper.py]: Put a line break after ':'.

File: lib/registration_helper.py [hunk 16]
```diff
@@ -0,0 +1,211 @@
...
+    cookies = CookieJar()
+    opener = urllib2.build_opener(urllib2.HTTPCookieProcessor(cookies))
+    opener.addheaders = [('Referer', RegistrationHelper.LOGIN_URL)]
+
+    try:
+      opener.open(RegistrationHelper.LOGIN_URL, urllib.urlencode(login_data))
+      return opener
+    except urllib2.HTTPError as error:
+      if error.code == cls.HTTP_UNAUTHORIZED:
+        print('Login failed. Please check your credentials and try again.')
```
Comment (menivaitsi) [inline lib/registration_helper.py]: Use appscale_logger methods for printing out to the console.

File: lib/registration_helper.py (hunk 15 above)
Comment (menivaitsi) [inline lib/registration_helper.py]: Do the same in the rest of the docstrings.

File: lib/registration_helper.py [hunk 17]
```diff
@@ -0,0 +1,211 @@
...
+    """ Try to retrieve a deployment ID from ZooKeeper to see if this
+    deployment has already been registered.
+
+    Args:
+      head_node: A string containing the IP address of the head node.
+      keyname: A string representing the SSH keypair name used for this
+        AppScale deployment.
+    Returns: A boolean indicating whether the deployment ID exists or not.
+    """
+    # Check if head node has a deployment id stored.
```
Comment (menivaitsi) [inline lib/registration_helper.py]: id -> ID

File: lib/registration_helper.py [hunk 18]
```diff
@@ -0,0 +1,211 @@
...
+
+  @classmethod
+  def select_project(cls, opener):
+    """ Asks the user to select a project to register the deployment under.
+
+    Args:
+      opener: A URL opener with valid cookies set for AppScale Portal access.
+      deployment_type: A string designating the type of deployment.
+      nodes: A list of containing the nodes layout.
+    """
```
Comment (menivaitsi) [inline lib/registration_helper.py]: Add a returns section here.

File: lib/registration_helper.py [hunk 19]
```diff
@@ -0,0 +1,211 @@
...
+
+  @classmethod
+  def register_deployment(cls, opener, deployment_type, nodes, project, name):
+    """ Asks the AppScale Portal for a new deployment ID.
+
+    Args:
+      opener: A URL opener with valid cookies set for AppScale Portal access.
+      deployment_type: A string designating the type of deployment.
+      nodes: A list of containing the nodes layout.
+    """
```
Comment (menivaitsi) [inline lib/registration_helper.py]: Add missing return section here.

**Comment by menivaitsi** [comment]
Have you run regression testing on this branch?

**Comment by menivaitsi** [comment]
:+1: 

Build passed.
Let's freeze this for now.

**Comment by nlake44** [comment]
Great work here guys! :dancer:

**Pull request merged**